File: rpn.fs
module RPN

let foldingFunction stack item =
  match (stack, item) with
  | (a::b::xs, "+") -> (b + a)::xs
  | (a::b::xs, "-") -> (b - a)::xs
  | (a::b::xs, "*") -> (b * a)::xs
  | (xs, numberString) -> (int numberString)::xs

(*
  let solveRPN = Array.fold foldingFunction List.empty<int> ("7 12 3 + *".Split(' '))
*)
let solveRPN (eq: string) = List.head <| Array.fold foldingFunction [] (eq.Split(' '))


[<EntryPoint>]
let main argv =
  printfn "should be 105: %d" (solveRPN "7 12 3 + *")
  printfn "should be -30: %d" (solveRPN "7 12 - 6 *")
  printfn "should be 32: %d" (solveRPN "6 2 + 5 3 - 2 * *")
  0

File: react/src/rpn.js
import List from "fable-core/List";
import { fold } from "fable-core/Seq";
import { split } from "fable-core/String";
export function foldingFunction(stack, item) {
  var matchValue = [stack, item];
  var $var1 = matchValue[0].tail != null ? matchValue[0].tail.tail != null ? matchValue[1] === "+" ? [0, matchValue[0].head, matchValue[0].tail.head, matchValue[0].tail.tail] : matchValue[1] === "-" ? [1, matchValue[0].head, matchValue[0].tail.head, matchValue[0].tail.tail] : matchValue[1] === "*" ? [2, matchValue[0].head, matchValue[0].tail.head, matchValue[0].tail.tail] : [3, matchValue[1], matchValue[0]] : [3, matchValue[1], matchValue[0]] : [3, matchValue[1], matchValue[0]];

  switch ($var1[0]) {
    case 0:
      return new List($var1[2] + $var1[1], $var1[3]);

    case 1:
      return new List($var1[2] - $var1[1], $var1[3]);

    case 2:
      return new List($var1[2] * $var1[1], $var1[3]);

    case 3:
      return new List(Number.parseInt($var1[1]), $var1[2]);
  }
}
export function solveRPN(eq) {
  return fold(function (stack, item) {
    return foldingFunction(stack, item);
  }, new List(), split(eq, " ")).head;
}